File: app/test/gamification/xp_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:gym_coach/gamification/xp_engine.dart';

void main() {
  test('workout completion awards 30 xp', () {
    expect(computeXpForWorkoutCompletion(allMealsCompleted: false).xp, 30);
  });

  test('workout + all meals awards 80 xp (30+50)', () {
    expect(computeXpForWorkoutCompletion(allMealsCompleted: true).xp, 80);
  });

  test('all 4 meals completed awards 20 xp', () {
    expect(
      computeXpForMealCompletion(mealsCompletedBefore: 3, totalMeals: 4).xp,
      20,
    );
  });

  test('partial meals awards 0 xp', () {
    expect(
      computeXpForMealCompletion(mealsCompletedBefore: 2, totalMeals: 4).xp,
      0,
    );
  });

  test('streak milestones', () {
    expect(computeXpForStreakMilestone(streakDays: 7).xp, 100);
    expect(computeXpForStreakMilestone(streakDays: 30).xp, 300);
    expect(computeXpForStreakMilestone(streakDays: 100).xp, 1000);
    expect(computeXpForStreakMilestone(streakDays: 365).xp, 5000);
    expect(computeXpForStreakMilestone(streakDays: 10).xp, 0);
  });

  test('badge unlock awards 200', () {
    expect(computeXpForBadgeUnlock().xp, 200);
  });

  test('coach adjustment awards 10', () {
    expect(computeXpForCoachAdjustment().xp, 10);
  });

  test('level math', () {
    expect(levelFromXp(0), 1);
    expect(levelFromXp(999), 1);
    expect(levelFromXp(1000), 2);
    expect(levelFromXp(2500), 3);
    expect(xpToNextLevel(0), 1000);
    expect(xpToNextLevel(1500), 500);
  });
}

File: app/android/app/src/test/java/com/gymcoach/gym_coach/gamification/XpEngineTest.kt
package com.gymcoach.gym_coach.gamification

import com.google.common.truth.Truth.assertThat
import org.junit.Test

class XpEngineTest {

    @Test
    fun `workout xp with all meals done is 80`() {
        assertThat(XpEngine.computeXpForWorkoutCompletion(allMealsCompleted = true).xp).isEqualTo(80)
    }

    @Test
    fun `workout xp with missing meals is 30`() {
        assertThat(XpEngine.computeXpForWorkoutCompletion(allMealsCompleted = false).xp).isEqualTo(30)
    }

    @Test
    fun `meal xp is 20 only on last remaining meal`() {
        assertThat(XpEngine.computeXpForMealCompletion(mealsCompletedBefore = 2, totalMeals = 3).xp).isEqualTo(20)
        assertThat(XpEngine.computeXpForMealCompletion(mealsCompletedBefore = 1, totalMeals = 3).xp).isEqualTo(0)
        assertThat(XpEngine.computeXpForMealCompletion(mealsCompletedBefore = 0, totalMeals = 0).xp).isEqualTo(0)
    }

    @Test
    fun `streak milestones match original fixed values`() {
        assertThat(XpEngine.computeXpForStreakMilestone(7).xp).isEqualTo(100)
        assertThat(XpEngine.computeXpForStreakMilestone(30).xp).isEqualTo(300)
        assertThat(XpEngine.computeXpForStreakMilestone(100).xp).isEqualTo(1000)
        assertThat(XpEngine.computeXpForStreakMilestone(365).xp).isEqualTo(5000)
        assertThat(XpEngine.computeXpForStreakMilestone(8).xp).isEqualTo(0)
    }

    @Test
    fun `badge and coach awards`() {
        assertThat(XpEngine.computeXpForBadgeUnlock().xp).isEqualTo(200)
        assertThat(XpEngine.computeXpForCoachAdjustment().xp).isEqualTo(10)
    }

    @Test
    fun `level is xp_div_1000_plus_1`() {
        assertThat(XpEngine.levelFromXp(0)).isEqualTo(1)
        assertThat(XpEngine.levelFromXp(999)).isEqualTo(1)
        assertThat(XpEngine.levelFromXp(1000)).isEqualTo(2)
        assertThat(XpEngine.levelFromXp(2500)).isEqualTo(3)
    }

    @Test
    fun `xpToNextLevel matches`() {
        assertThat(XpEngine.xpToNextLevel(0)).isEqualTo(1000)
        assertThat(XpEngine.xpToNextLevel(1500)).isEqualTo(500)
    }
}
